### ISSUE TITLE
match-define does not propagate type annotations to be used

### ISSUE THREAD

**Issue by mfelleisen** [racket/typed-racket#829]
### What version of Racket are you using?

7.3.03 May 2019

### What program did you run?

```
#lang typed/racket

(struct circle ({x : Integer} {y : Integer}) #:transparent)

(define-type ClassType (Class (init-field {c circle})))

(define c : ClassType
  (class object% (init-field c)
    (match-define (circle #{x : Integer} #{y : Integer}) c)
    (displayln `(,x + ,y = ,(+ x y)))
    (super-new)))
```

### What should have happened?

It should type check. 

### If you got an error message, please include it here.

It signals type errors: 

```
Type Checker: type mismatch
  expected: Number
  given: Any in: x
. Type Checker: type mismatch
  expected: Number
  given: Any in: y
. Type Checker: type mismatch
  expected: Any
  given: (List Any '+ Any '= Nothing) in: (quasiquote ((unquote x) + (unquote y) = (unquote (+ x y))))
. Type Checker: Summary: 3 errors encountered in:
  x
  y
  (quasiquote ((unquote x) + (unquote y) = (unquote (+ x y))))
```

### Workaround 

Specifying the types explicitly, outside of `match-define` works: 

```
(define c : ClassType
  (class object% (init-field c)
    (: x Integer)
    (: y Integer)
    (match-define (circle x y) c)
    (displayln `(,x + ,y = ,(+ x y)))
    (super-new)))
```

**Comment by AlexKnauth** [comment]
I don't think it's a problem with `match` or `match-define`, it's a problem with `class` not dealing with `#{id : type}` annotations properly in `define-values` declarations.

```racket
#lang typed/racket

(struct circle ({x : Integer} {y : Integer}) #:transparent)

(: f : circle -> (values Integer Integer))
(define (f c) (values 1 2))

(define-type ClassType (Class (init-field {c circle})))

(define c : ClassType
  (class object% (init-field c)
    (define-values [#{x : Integer} #{y : Integer}] (f c))
    (displayln `(,x + ,y = ,(+ x y)))
    (super-new)))
```

Again a workaround is to put `(: x Integer)` and `(: y Integer)` declarations outside the `define-values` form.

The issue is that the `#{x : Integer}` and `#{y : Integer}` annotations inside the `define-values` form are ignored / thrown away by the `class` form.

**Comment by samth** [comment]
That means that the `typed/racket` version of the `class` macro isn't looking at the `'type-label` syntax property added by `#{x : T}`. Maybe @takikawa has a suggestion for where the fix should be?

**Comment by takikawa** [comment]
I might try to fix this later if I have time, but I think the easiest solution would be to have the class macro in `base-env/class-prims.rkt` process the `'(declare-field-initialization identifier)` declarations that the regular class macro inserts and then check if `identifier` has a `'type-label` or not. If it does, then use that in the class typechecker's type annotation table.

(and I guess like TR without classes, it should signal a duplicate annotation error if you provided both a top-level or class-macro type annotation and the type label one)

**Comment by takikawa** [comment]
Closed by c8b119085c7838b2b3e9c5d7765c74f0ec6548a3
